From 3f4be5867d1b2b7db708fdd2e5c13885c2dcf056 Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:26:18 -0400
Subject: [md] don't allow bitmap file to be added to raid0/linear

Message-id: <1409563584-9000-77-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 90017
O-Subject: [RHEL7.1 PATCH 76/82] md: don't allow bitmap file to be added to raid0/linear.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

An array can only accept a bitmap if it will call bitmap_daemon_work
periodically, which means it needs a thread running.

If there is no thread, don't allow a bitmap to be added.

Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit d66b1b395a59027a1c054e1fc57d582cb4194491)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/md.c b/drivers/md/md.c
index 4749c8a..1204ed6 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -5961,7 +5961,7 @@ static int set_bitmap_file(struct mddev *mddev, int fd)
  int err = 0;
 
  if (mddev->pers) {
-  if (!mddev->pers->quiesce)
+  if (!mddev->pers->quiesce || !mddev->thread)
    return -EBUSY;
   if (mddev->recovery || mddev->sync_thread)
    return -EBUSY;
@@ -6263,7 +6263,7 @@ static int update_array_info(struct mddev *mddev, mdu_array_info_t *info)
   rv = update_raid_disks(mddev, info->raid_disks);
 
  if ((state ^ info->state) & (1<<MD_SB_BITMAP_PRESENT)) {
-  if (mddev->pers->quiesce == NULL)
+  if (mddev->pers->quiesce == NULL || mddev->thread == NULL)
    return -EINVAL;
   if (mddev->recovery || mddev->sync_thread)
    return -EBUSY;
-- 
1.7.1